hbase-it/src/test/java/org/apache/hadoop/hbase/CoprocClusterManager.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase;
+
+import java.io.IOException;
+import java.util.Objects;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hbase.client.AsyncAdmin;
+import org.apache.hadoop.hbase.client.AsyncConnection;
+import org.apache.hadoop.hbase.client.ConnectionFactory;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecRequest;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecResponse;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecService;
+import org.apache.hadoop.hbase.util.Pair;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Overrides commands to make use of Coproc where possible.
+ */
+@InterfaceAudience.Private
+public class CoprocClusterManager extends HBaseClusterManager {
+  private static final Logger LOG = LoggerFactory.getLogger(CoprocClusterManager.class);
+
+  @Override
+  protected Pair<Integer, String> exec(String hostname, ServiceType service, String... cmd)
+    throws IOException {
+    // we don't have ssh to log in as the service user, so wrap the command in sudo.
+    // assumes user running the hbase process has sudo privileges to target service.
+    final String serviceUser = getServiceUser(service);
+    final String[] commandPrefix =
+      !Objects.equals(serviceUser, getServiceUser(ServiceType.HBASE_MASTER))
+        ? new String[] { "sudo", "-u", serviceUser }
+        : new String[0];
+    final String command = StringUtils.join(ArrayUtils.addAll(commandPrefix, cmd), " ");
+    LOG.info("Executing remote command: {}, hostname:{}", command, hostname);
+
+    try (final AsyncConnection conn = ConnectionFactory.createAsyncConnection(getConf()).join()) {
+      final AsyncAdmin admin = conn.getAdmin();
+      final ShellExecRequest req = ShellExecRequest.newBuilder()
+        .setCommand(command)
+        .build();
+
+      final ShellExecResponse resp;
+      switch(service) {
+        case HBASE_MASTER:
+          // TODO: what happens if the intended action was killing a backup master?
+          resp = masterExec(admin, req);
+          break;
+        case HBASE_REGIONSERVER:
+          final ServerName targetHost = resolveRegionServerName(admin, hostname);
+          resp = regionServerExec(admin, req, targetHost);
+          break;
+        default:
+          throw unsupportedServiceType(service);
+      }
+
+      LOG.info("Executed remote command, exit code:{} , output:{}", resp.getExitCode(),
```

Comment:
Depends on if you log the thread name as well. Sure, it doesn't hurt. Although, since we call the async method, there's no response code or output available. Should just truncate the log message.